proof/refine/ARM/Reply_R.thy

```diff
@@ -8,31 +8,28 @@ theory Reply_R
 imports TcbAcc_R
 begin
 
-crunches setReplyTCB for
-  pred_tcb_at'[wp]: "pred_tcb_at' proj test t"
+crunches setReplyTCB
+  for pred_tcb_at'[wp]: "\<lambda>s. P (pred_tcb_at' proj test t s)"
+  and tcb_at'[wp]: "\<lambda>s. P (tcb_at' t s)"
 
-lemma replyUnlink_st_tcb_at'_wp:
-  "\<lbrace>\<lambda>s. reply_at' rptr s \<longrightarrow>
-          obj_at' (\<lambda>reply.
-                      (replyTCB reply = Some tptr \<longrightarrow> test Inactive) \<and>
-                      (replyTCB reply \<noteq> Some tptr \<longrightarrow> st_tcb_at' test tptr s))
-                  rptr s\<rbrace>
-   replyUnlink rptr
-   \<lbrace>\<lambda>_. st_tcb_at' test tptr\<rbrace>"
+lemma replyUnlink_st_tcb_at':
+  "\<lbrace>\<lambda>s. tcb_at' t s \<longrightarrow> (if t'=t then P (P' Inactive) else P (st_tcb_at' P' t' s))\<rbrace>
+    replyUnlink r t
+   \<lbrace>\<lambda>rv s. P (st_tcb_at' P' t' s)\<rbrace>"
   unfolding replyUnlink_def
-  supply if_split[split del]
-  apply (wpsimp wp: sts_st_tcb' hoare_vcg_if_lift2 hoare_vcg_imp_lift
-                    hoare_vcg_disj_lift threadSet_pred_tcb_at_state gts_wp'
-              simp: getReplyTCB_def cong: conj_cong disj_cong)
-  by (fastforce simp: obj_at'_def)
+  apply (wpsimp simp: getReplyTCB_def
+                  wp: sts_st_tcb_at'_cases_strong gts_wp' hoare_vcg_imp_lift
+                cong: conj_cong split: if_split_asm)
+done
 
 lemma replyUnlink_st_tcb_at'_Inactive:
   "\<lbrace>\<lambda>s. st_tcb_at' test tptr s \<and> test Inactive\<rbrace>
-   replyUnlink rptr
+   replyUnlink rptr tptr'
    \<lbrace>\<lambda>_. st_tcb_at' test tptr\<rbrace>"
-  apply (wpsimp wp: replyUnlink_st_tcb_at'_wp)
+  apply (wpsimp wp: replyUnlink_st_tcb_at')
   done
 
+(* Not used
```

Comment:
Maybe remove this.

Reply:
Well, actually, I think it makes sense to leave these commented out. The changes here will affect the ongoing proofs, and depending on how that goes, these might not be needed at all or there could be some use for them. We might want to wait.

Reply:
Would it make sense to leave the statement in, but oops the lemma? Might make it slightly easier to find and at least be type checked (although it's not going to turn up in `find_theorems`, so maybe not that important).

In any case, commented-out blocks should get a `FIXME RT` so that we remember to remove them if we didn't need them in the end.

Reply:
Yes, I've talked to Michael and will update those with more info.